### ISSUE TITLE
Token migration support

### ISSUE THREAD

**Issue by maraoz** [OpenZeppelin/openzeppelin-solidity#148]
See Golem's token migration system: 
https://github.com/golemfactory/golem-crowdfunding/blob/50100b27a7c6841ed430a028d100f5d45ba08fb1/contracts/Token.sol#L85

**Comment by DavidKnott** [comment]
@maraoz This feature sounds really useful and interesting!  Is the implementation you're thinking of a ```MigrateToken``` that inherits from ```StandardToken``` or adding on to ```StandardToken``` itself?

**Comment by maraoz** [comment]
`MigrateToken` sounds like a better idea, we want to modularize as much as possible.

**Comment by DavidKnott** [comment]
@maraoz Okay I'll get started on it!

**Comment by elopio** [comment]
From the [Golem token whitepaper](https://golem.network/doc/GNTterms.pdf):

>Migrate token - a function which implements GNT migration to another contract.
>Does nothing by default, but if at some point a  GNT upgrade is required, a separate migration contract can be specified and recommended by Golem Factory GmbH to be used to transfer tokens to the new contract.
>Technically speaking, if a GNT upgrade is required, a contract implementing the ​MigrationAgent interface is created and set by Golem Factory GmbH in the GNT contract (for security reasons this can be done only once). Following that, each GNT holder can decide whether to call  MigrationAgent.migrateTokens to transfer GNT to the new contract, or not.
>
>MigrationAgent can be implemented only after the new token is implemented and deployed. That’s why only an interface is provided right now. Migration is to be used if it turns out at some point that token upgrade is needed for whatever reason (e.g. changes in Ethereum, or changes in Golem’s
design). The upgrade will need action from token holders and cannot be imposed by Golem Factory GmbH.

I'm removing the research tag because it's clear what to do.

It would be interesting to see what of these things will no longer be necessary with the upgradeability of zeppelinOS. But anyway, it would be nice to have in openzeppelin as a building block for custom solutions.

@DavidKnott are you still interested in implementing this?

**Comment by nventuro** [comment]
We've had [`TokenMigrator`](https://github.com/OpenZeppelin/openzeppelin-solidity/blob/v2.2.0-rc.1/contracts/drafts/ERC20Migrator.sol) for a while now.
